src/diffusers/loaders/lora.py

```diff
@@ -1033,17 +1036,71 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
     def set_adapters(
         self,
         adapter_names: Union[List[str], str],
-        adapter_weights: Optional[List[float]] = None,
+        adapter_weights: Optional[Union[List[float], float, List[Dict], Dict]] = None,
     ):
+        adapter_names = [adapter_names] if isinstance(adapter_names, str) else adapter_names
+
+        allowed_numeric_dtypes = (float, int)
+        has_second_text_encoder = hasattr(self, "text_encoder_2")
+
+        # Expand weights into a list, one entry per adapter
+        if adapter_weights is None or isinstance(adapter_weights, (allowed_numeric_dtypes, dict)):
+            adapter_weights = [adapter_weights] * len(adapter_names)
+
+        if len(adapter_names) != len(adapter_weights):
+            raise ValueError(
+                f"Length of adapter names {len(adapter_names)} is not equal to the length of the weights {len(adapter_weights)}"
+            )
+
+        # Normalize into dicts
+        allowed_keys = ["text_encoder", "unet"]
+        if has_second_text_encoder:
+            allowed_keys.append("text_encoder_2")
+
+        def ensure_is_dict(weight):
+            if isinstance(weight, dict):
+                return weight
+            elif isinstance(weight, allowed_numeric_dtypes):
+                return {key: weight for key in allowed_keys}
+            elif weight is None:
+                return {key: 1.0 for key in allowed_keys}
+            else:
+                raise RuntimeError(f"lora weight has wrong type {type(weight)}.")
+
+        adapter_weights = [ensure_is_dict(weight) for weight in adapter_weights]
+
+        for weights in adapter_weights:
+            for k in weights.keys():
+                if k not in allowed_keys:
+                    raise ValueError(
+                        f"Got invalid key '{k}' in lora weight dict. Allowed keys are 'text_encoder', 'text_encoder_2', 'down', 'mid', 'up'."
+                    )
+
+        # Decompose weights into weights for unet, text_encoder and text_encoder_2
+        unet_weights, text_encoder_weights = [], []
+        if has_second_text_encoder:
+            text_encoder_2_weights = []
+
+        for weights in adapter_weights:
+            unet_weight = weights.get("unet", None)
+            text_encoder_weight = weights.get("text_encoder", None)
+            if has_second_text_encoder:
+                text_encoder_2_weight = weights.get("text_encoder_2", None)
+
+            unet_weights.append(unet_weight)
+            text_encoder_weights.append(text_encoder_weight)
+            if has_second_text_encoder:
+                text_encoder_2_weights.append(text_encoder_2_weight)
```

Comment:
```suggestion
        for adapter_name, weights in zip(adapter_names, adapter_weights):
            if isinstance(weights, dict):
                unet_weight = weights.pop("unet", None)]
                text_encoder_weight = weights.pop("text_encoder", None)
                text_encoder_2_weight = weights.pop("text_encoder_2", None)
                if len(weights) >0:
                    raise ValueError(f"Got invalid key '{weights.keys()}' in lora weight dict for adapter {adapter_name}.")
               if text_encoder_2_weight is not None and not hasattr(self, "text_encoder_2"):
                   logger.warning("lora weight dict contains text_encoder_2 weights but will be ignored because pipeline does not have text_encoder_2")
            else:
               unet_weight = weights
               text_encoder_weight = weights
               text_encoder_2_weight = weights
       
            unet_weights.append(unet_weight)
            text_encoder_weights.append(text_encoder_weight)
            text_encoder_2_weights.append(text_encoder_2_weight)
```